[mono-move] Design doc for native function interface (#19856)
aptos-labs/aptos-core: third_party/move/mono-move/docs/native_functions_interface.md

# Native Function Interface — Draft

> **Status:** Informal first draft. Companion: [`native_functions_existing.md`](native_functions_existing.md) — survey of the current Aptos VM native interface.
>
> **Note:** Rust signatures throughout this doc are illustrative — exact shapes may be adjusted during implementation.

## 0. Goals and Milestone

> **Goals.** Native functions are first-class citizens in MonoMove — direct access to VM internals (stack and heap memory, the loader, the gas meter, etc.), behaving conceptually like VM instructions rather than opaque external calls, subject to the same safety and metering guarantees as the rest of the VM.

Reviewer: ```suggestion
**Goals.** Native functions are first-class citizens in MonoMove — direct access to VM internals (stack and heap memory, the loader, the gas meter, etc.), behaving conceptually like VM instructions rather than opaque external calls, subject to the same safety and metering requirements as the rest of the VM.
```

Immediate milestone: **prototype that can run a Decibel transaction end-to-end** (see [`native_functions_decibel_minimum.md`](native_functions_decibel_minimum.md) for the ~40-native MVP set). Long-term goals (typed argument access, full memory safety) are flagged and punted.

---

## 1. Registration & Resolution

> - **Specializer is totally ignorant of native bodies.** It reads only the function handles declared in `CompiledModule`. For a `native`-marked handle it emits a generic call site — no callee monomorphization, type args passed at runtime. It never inspects the Rust impl.

Reviewer: - Briefly also mention closures over natives.
- "totally ignorant" -> agnostic.

> - **Linking is lazy** — natives bind to call sites at first invocation, not at module load. Matches existing VM behavior; modules that declare unimplemented natives still load.

Reviewer: Maybe we should say something about whether the lazy linking can fail at runtime (looks like it can, given what's written here) and what happens in that case.

Author: Yes, this is exactly how I intend it to work. Two benefits:
1. Compatibility with the current behavior (runtime linking failure)
2. Making it easy for us to test unimplemented native functions

- Natives register with the runtime as `(addr, module, name) → NativeImpl`. `NativeImpl` = Rust fn pointer + declared ABI.
- At dispatch the `Function` struct is either a Move body or a native body. Call/return path is uniform per [`stack_and_calling_convention.md`](stack_and_calling_convention.md).
- **Selective monomorphization** (specializing a generic native per type-arg) is **out of scope for now**. It's the one place the specializer would need richer per-native knowledge.

---

## 2. Native Function Signature & Context

> `NativeContext` is a **trait**, not a concrete type. A native is written generic over it:

Reviewer: ```suggestion
`NativeContext` is a **trait**. A native method is generic over this trait:
```

```rust
trait NativeContext { /* methods listed below */ }

fn my_native<C: NativeContext>(ctx: &mut C) -> NativeResult { ... }
```

> Each VM instance commits to one concrete `C` (`ProductionContext`, `TestContext`, etc.). The function table holds `type NativeFunction<C> = Box<dyn Fn(&mut C) -> NativeResult>` with `C` fixed, so every context call statically dispatches.

Reviewer: "the function table" => unclear from just this context what you mean.

Author: Let me rephrase this. Here what I meant is that a VM instance holds a collection/list of native functions over some concrete context type `C`.

> Because the native is a `Fn` closure, it can carry its own captured state — useful for small per-native config or constants. For anything that needs to persist across calls or be shared between natives, use the extension system (§6) instead.

Reviewer 2: But we have context which can capture the state? Do you have an example where this is useful (we also have `ctx` which stores configs)

Author: I was originally thinking about gas parameters but guess those can be carried by the context indeed. (Though doing it this way requires natives to be directly aware of the "path" at which its gas parameters are stored -- not a major issue though.)

Let me think a little bit more about this -- the motivation I had was that (1) extensions are somewhat slower due to the type-id lookup (2) the cost of an indirect call (through the vtable) is likely negligible, so why not having a bit more flexibility here.

Context APIs (tentative):

```rust
ctx.num_args() -> usize
ctx.arg<T: VMValue>(i: usize) -> T

ctx.num_returns() -> usize
ctx.set_return<T: VMValue>(i: usize, val: T)
```

`VMValue` is implemented for primitives (`u8`, `u64`, `bool`, `AccountAddress`, ...) and for VM type wrappers around common heap data structures (struct pointer, vector, reference). Adding support for a new ABI-passable type is one `impl VMValue for ...` away.

The `i` is a positional index, not a raw frame offset — the context resolves it against the native's declared ABI on each call. The indirection is for safety: natives can't walk into the wrong slot. If the lookup ever becomes a hot-path concern, an unsafe `arg_at_offset` escape hatch is possible.

- **Static ABI cross-check at module load.** Rust natives declare their ABI at registration; the loader compares against the `native fun` declaration in `CompiledModule` and fails the load on mismatch. Catches Rust/Move drift at deploy time. Deliberate exception to lazy linking (§1).

**Open questions:**

> - **Aliasing across repeated `arg` calls.** `arg::<VectorPtr>(0)` called twice returns two wrappers pointing at the same heap chunk — aliased, not copied. Resolution TBD.

Reviewer 2: natives are inherently dangerous unsafe code, I think this is fine: we can have LLMs catching such bugs and tests

Author: Yeah should probably make it more explicit that native function safety is inherently a "best effort" -- we'll do what can be cheaply done, but ultimately natives need to have direct access to the VM internals if they need to and are responsible for their own correctness.

- **No shadow arenas.** Today's `AlgebraContext.objs` / `NativeRistrettoPointContext.points` aren't carried forward — uncounted growth, beyond the gas budget. The crypto natives must be **rewritten**: the handle-into-arena indirection (`RistrettoPoint { handle: u64 }` + Rust-side `Vec`) becomes a real heap struct holding the bytes directly, with a proper `ObjectDescriptor`. Off the Decibel path; separate milestone.
- Transient Rust-side scratch inside a single call is fine. Rule: anything that becomes a Move value, or grows with input size, lives on the VM heap.

---

## 5. Type Reflection

`InternedType` (defined in `core/src/types.rs`) is the VM's runtime type handle. It's **directly traversable**: natives can walk a type in place — kind, fields, layout, element type, etc. — without ever materializing a separate `TypeTag` or `MoveTypeLayout`.

> When a tag or layout actually IS needed (e.g. embedded in an event header, or serialized for external consumption), they can be derived on demand. The goal is to make materialization the exception, not every reflection-using native's default.

Reviewer 2: When it is actually needed? I don't think we need it at all?

Author: I don't know just yet, so just being conservative and leaving the doors open here. If it turns out we don't need it, then it's even better.

Context APIs (tentative):

```rust
ctx.num_ty_args() -> usize
ctx.ty_arg(i) -> InternedType

// Materialized escape hatches
ctx.ty_to_tag(ty) -> TypeTag
ctx.ty_to_layout(ty) -> MoveTypeLayout
```

Type-structural lookups (kind, fields, element type, size/align, etc.) ideally live as methods on `InternedType` itself (`ty.kind()`, `ty.fields()`, ...), so natives can inspect types without routing through the context. If that doesn't pan out — borrow-checker or other constraints — the fallback is `ctx.ty_kind(ty)`-style methods on the context.

**Dispatching on a known struct.** The runtime type representation carries enough information to support this — which is what `crypto_algebra::*` relies on today, in the absence of native specialization (§1).

---

## 6. Native Extensions

Some natives need state that persists across many calls within a single transaction — counters, accumulators, caches, custom data structures. Native extensions are the mechanism for this.

> **Shape.** An extension is a Rust-side struct registered with the VM at startup, instantiated fresh per session. It can carry both:

Reviewer 2: I think instead we should have it reset per session, for current VM we have SessionListener API (which never really got implemented for sessions lol) - we need to mimic similar things. If your extension has some carried over state (tables, aggregators) it does not need to re-initialise but it may need to update session ID

Author: Hmm... are you thinking about preserving extension states across sessions? Not sure how that'll work as a key pattern many extensions do is to allocate their own memory on the heap, which has the lifetime of a session, except for the global storage data, which is kept and frozen after compaction -- maybe that's what you are thinking about?

Reviewer 2: But we need to preserve extension states already for tables, events, aggregators? Also `UserTransactionContext` or others can be persisted, they were not for historical reasons - but I think this is something we can evaluate and consider? 

But heap's lifetime is not session, its transaction? Global values live there so you need to persist + it exactly bounds what a single txn can do?

Author: There is some confusion over the terminology here. When I said "session" I was referring to the "revertible session" we currently want to implement in the new VM, which does have the same lifetime as the transaction itself.

Now when it comes to native extensions, you are right that some may need to have their states updated mid-transaction. The semantics though, is less clear at the moment. I'm not sure a "universal reset" applies to all extensions -- each may want to handle it differently (especially for roll-backs).

Reviewer 2: "each may want to handle it differently" - this works, I think as long as we have a way to make sure implementations are also safe wrt to rollback, etc. and make sure that we minimise resets (currently we drop all data from extensions and create them again - we can also just clear vectors, etc. to avoid drops - this is what I meant mostly)

```rust
type ExtensionFactory = fn(&mut Heap) -> Box<dyn NativeExtension>;
```

At session start the VM invokes each factory and stows the resulting `Box<dyn NativeExtension>` in a `HashMap<TypeId, Box<dyn NativeExtension>>`. `get_extension::<T>` does the `TypeId` lookup + downcast.

> **Alternative (Option C, not chosen): typed registry.** Skip the trait-object map entirely and have the VM hold a struct with one named field per extension type (`Extensions { event_store: EventStore, transaction_context: TransactionContext, … }`). `get_extension::<T>()` becomes a type-dispatched field accessor. Lets the trait keep `init -> Self` and `impl Iterator` (RPITIT), and removes dyn dispatch. Trade-off: third parties can't add extensions without modifying the registry struct. Worth revisiting if MonoMove never grows out-of-tree extensions.

**Full list of extensions in MonoMove** (after the rework — current VM has 10, listed in `aptos-vm/src/move_vm_ext/session/mod.rs::make_aptos_extensions`):

- **EventStore** ([`event_store.md`](event_store.md)) — heap-allocated container with checkpoint/rollback.

> - **TransactionContext** — Rust-side only (AUID counter, session counter, optional `UserTransactionContext`).

Reviewer 2: Where is UserTransactionContext allocated?

Author: Hopefully on the rust side (just like what it is today). Not ideal as it's a big struct, but it's probably safe as it's still somewhat bounded?

Reviewer 2: We can start with Rust side I guess and then move more things to same memory heap area 👍

- **RandomnessContext** — Rust-side only (8-byte counter + flag).

> - **CodeContext** — small Rust-side wrapper holding an optional pointer to a heap-allocated publish request (which can be large — module bundles + metadata).

Reviewer 2: Hm does this mean modules to publish are part of heap? Packages can be 1-1.8Mb in size, is this a concern?

Author: > Hm does this mean modules to publish are part of heap?

Yep, this is exactly what is desired -- this additional space taken should count towards a transaction's memory limit. We can bump the maximum memory limit if 10 MB turns out to be too restrictive, after covering thing like this.

There's another potential advantage of keeping this in the heap: we may no longer need to convert these from Move values into Rust, and could end up saving both some space and computation.

- **StateStorageContext** — Rust-side only (borrowed host resolver).

> - **ObjectContext** — memo cache for `create_user_derived_object_address_impl`. Pure compute optimization; could be dropped, or if kept should live on the heap (same accounting argument as the crypto arenas, §4).

Reviewer 2: We cannot drop it - it's the cache? Why it should be on the heap? It's a pure compute optimization we can keep in actual memory through txn lifetime?

Author: Well, the current implementation has this cache tied to the lifetime of a transaction:
```
#[derive(Default, Tid)]
pub struct NativeObjectContext {
    // TODO - if further optimizations is important, we can consider if:
    //   - caching all (or just some derive_from) locations is useful
    //   - if it is faster to use BTreeMap or HashMap, given the lenghts of the addresses
    //   - if it is worth moving to native/caching other address deriving as well
    derived_from_object_addresses:
        RefCell<HashMap<(AccountAddress, AccountAddress), AccountAddress>>,
}
```
Honestly it's not immediately clear whether we really need this cache -- sure the computation here may be expensive, but are we even calling with the same arguments more than once in the same transaction?

Reviewer 2: I think it was added because of some perf SEV (calling this in a loop or similar)

> - **AlgebraContext / RistrettoContext** — after the crypto-native rewrite (§4), the natives operate on heap-allocated curve points directly; these extensions may shrink to gas/feature-flag plumbing or go away entirely.

Reviewer 2: I don't think I understand what this sentence here means

Author: These are crypto natives that currently allocate mathematical objects through their own shadow memory. Was just saying that those need to be re-implemented using the runtime's heap. Let me make the wording here a bit clearer.

Reviewer 2: Yes, I think "these extensions may shrink to gas/feature-flag plumbing or go away entirely." part what was confusing as this part does not make much sense 😄

> - **TableContext, AggregatorContext** — core functionality (table storage, aggregator / delayed-field bookkeeping) moves into the runtime's global value system (§7). What remains, if anything, is a thin shim translating native calls into runtime storage operations; the extension may go away entirely.

Reviewer 2: Interesting! So you see these being part of VM so extension not needed? 

We can make them part of global storage and it's probably better to jeep all reads/writes together instead of trying to split them unnaturally across extensions or not. Note that access is different for them though:
- Tables: need hash of txn per session, has its own key, so we need table resolver trait or similar in any way?
- Aggregators: lookup by ID (but as discussed let's start with everything in memory and conflicting) 
Both need rollback and checkpoints.

Author: Yes, tables and aggregators are low-level storage primitives, better to be managed together with other global values. Not all details are clear at this moment -- let's see what it takes to implement them.

**Open question: rollback support.** Context: MonoMove is migrating away from the current VM's multi-session model (separate prologue / user / epilogue sessions stitched via `RespawnedSession`) toward a **single session** spanning the whole transaction, with rollback for partial failures. So extensions are created once and need to survive all three phases; rollback happens in-place. Two ways:

- *Single set + per-extension rollback:* each extension implements checkpoint/rollback only if it needs to. EventStore already does; counters / memo caches / monotonically-increasing things don't. Aligned with the single-session direction.

> - *Spawn fresh per sub-session:* a fresh set per sub-phase, merged across boundaries. Matches today's Aptos VM, but cuts against the single-session goal.

Reviewer 2: I think this is a no-go?

Author: I think so -- "but cuts against the single-session goal".

Settle once the sub-session model is pinned down.

---

## 7. Global Value System

- **Delegate entirely.** The native interface exposes thin operations that pass through to the global value system (direction at #19536; not finalized).
- Checkpoint / rollback is the global value system's concern, not the native interface's.

Context APIs (tentative):

```rust

> ctx.exists_at(addr: AccountAddress, ty: InternedType) -> bool

Reviewer 2: I think we only have existence native but we can delegate yes

Author: Good point, we may not need these other global storage APIs just yet.

We'll likely need the ones for table though.

Reviewer 2: We need to tweak API then to have some `StorageKey`

---

## 8. Events

Delegated to the event store ([`event_store.md`](event_store.md)). Natives only see emit; checkpoint and rollback are driven by the execution engine.

Context API:

```rust
ctx.emit_event(ty: InternedType, value: ValuePtr) -> Result<(), _>
```

`ty` and `value` come from the calling Move frame via §3's accessors; the context plumbs them through to the event store.

---

## 9. Results / Errors

All errors are transaction-aborting — execution state is discarded on failure, so natives don't need to clean up the frame or restore state. On success, the native must have written the correct number of return values at the expected offsets; the interpreter proceeds unconditionally.

**Status: TBD.** Two open questions:

1. **`Result<(), NativeError>` vs. flat enum** (`NativeResult { Success, Abort, OutOfGas, ... }`). Flat avoids the `Result` discriminant on the success path.
2. **Embed the VM's internal error type, or stay parallel?** Embedding makes propagation from sub-systems trivial; parallel keeps the native ABI decoupled.

Variants needed regardless: user abort with code, out-of-gas / limit-exceeded, propagated sub-system error, invariant violation.

> Lands once the broader VM error story in [`error_design.md`](error_design.md) settles.

Reviewer 2: This has settled?

Author: Yes, not a blocker for the milestone, so I haven't fleshed this part out.

Discussed with @calintat as well and would love to have him chime in on the exact shape of the native result here.

---

## 10. Code & Function

**TODO.** All deferred:

- Function resolution & loading
- Function values / closures
- Dynamic dispatch (e.g. dispatchable fungible assets, account abstraction)

> - Module traversal

Reviewer 2: We definitely do not need traversal it was only before lazy loading

- Caller-frame introspection (e.g. `event::write_module_event_to_store` reading the caller's module ID for the cross-module-emit check)

---

## 11. Gas Metering

**TODO.** Detailed design lives in [`gas_design.md`](gas_design.md). High-level direction for natives:

- **Constant-cost natives** (e.g. `signer::borrow_address`) — fixed cost, charge upfront. Could potentially batch with the surrounding basic block's gas charge, though savings may not justify the complexity.
- **Data-dependent cost** (vector operations, serialization, etc.) — inspect input size, charge upfront before the work.
- **Iterative or unpredictable cost** — charge incrementally (e.g. per loop iteration). Per-iteration charging still satisfies charge-before-work. Post-hoc charging is acceptable when cost is most naturally computed after the work, but only for transient bounded-constant violations that are carefully reviewed.

Contract: **charge before work**. See [`vm_security_and_correctness.md`](vm_security_and_correctness.md) for broader gas-metering invariants.

---

## 12. Miscellaneous Open Items

> - **Rayon isolation for crypto natives.** Today's pairing and multi-scalar-multiplication natives wrap their work in `with_native_rayon(...)` to avoid deadlocking against the block executor's own rayon pool (the underlying crypto crates use `rayon::par_iter` internally). MonoMove will need an equivalent mechanism.

Reviewer 2: Not sure let's check with @zekun000 - iirc this was hack to fix a deadlock which did not work, so rayon was removed from executor, so this can be removed/ignored
